File: moonPos.py
from jdCalc import julianCalc
import json 
import math

#override sin and cos function that accept degree
def toRadian(degree):
    return degree*math.pi/180

sin = lambda x: math.sin(toRadian(x))
cos = lambda x: math.cos(toRadian(x))


def reduceAngle(func):
    """A decorator that processes that reduce large angles into less than 360#"""
    def wrapper(self, T):
        y = func(self, T)
        return y%360
    return wrapper
    
class moonPosition:
    def __init__(self):
        self.longitude, self.latitude, self.distance = self.import_moon_coef()
    
    def import_moon_coef(self):
        with open('coefficients.json', 'r') as file:
            data = json.load(file)
        longitude = data['moonPosition']['longitude']
        latitude = data['moonPosition']['latitude']
        distance = data['moonPosition']['distance']
        return longitude, latitude, distance
        
    @reduceAngle
    def mean_longitude(self, T):
        L = (218.3164477 + 481267.88123431*T - 0.0015786*T**2 + 
        T**3/538841 - T**4/65194000)
        return L 
    
    @reduceAngle
    def mean_elongation(self, T):
        D = (297.8501921 + 445267.1114034*T - 0.0018819*T**2
        +T**3/545868 - T**4/113065000)
        return D 
    
    @reduceAngle
    def sun_mean_anomaly(self, T):
        M = (357.5291092 + 35999.0502909*T - 0.0001536*T**2 + T**3/24490000)
        return M 
    
    @reduceAngle 
    def moon_mean_anomaly(self, T):
        m2 = (134.9633964+ 477198.8675055*T+ 0.0087414*T**2+ T**3/69699- T**4/14712000)
        return m2
    
    @reduceAngle
    def moon_latitude_arguments(self, T):
        F = 93.2720950 + 483202.0175233*T - 0.0036539*T**2 - T**3/3526000 + T**4/863310000
        return F 
    
    @reduceAngle
    def A1_func(self, T):
        A1  = 119.75+131.849*T
        return A1 
    
    @reduceAngle
    def A2_func(self, T):
        A2 = 53.09+ 479264.290*T
        return A2 
    
    @reduceAngle
    def A3_func(self, T):
        A3 = 313.45 + 481266.484*T 
        return A3
    
    def compute_E(self, T):
        E = 1 - 0.002516*T - 0.0000074*T 
        return E 
    
    def _compute_coef(self, coef_dataset, coef_number = None, use_sin = True, **kwargs):
        '''Computing Moon Position Coefficient'''
        
        D = kwargs['D']
        M = kwargs['M']
        m2 = kwargs['m2']
        F = kwargs['F']
        E = kwargs['E']
        
        sum_all = 0
        terms = coef_number if coef_number else len(coef_dataset)
        func = sin if use_sin else cos
        
        for coef in coef_dataset[:terms]:
            param = coef[0]*D + coef[1]*M + coef[2]*m2 + coef[3]*F
            one_term = coef[4]*func(param)
            if abs(coef[1]) == 1:
                one_term*=E
            elif abs(coef[1]) == 2:
                one_term*=E**2
            sum_all += one_term
        
        return sum_all
    
    def compute_long_cor(self, coef_number = None, cor_else = True, **kwargs):
        
        A1 = kwargs['A1']
        A2 = kwargs['A2']
        F = kwargs['F']
        L = kwargs['L']
        long_cor = self._compute_coef(self.longitude, coef_number, **kwargs)
        
        if cor_else:
            long_cor += 3958*sin(A1)+ 1962*sin(L-F)+ 318*sin(A2)
        
        return long_cor
    
    def compute_dist_cor(self, coef_number = None, **kwargs):
        dist_cor = self._compute_coef(self.distance, coef_number, use_sin=False, **kwargs)
        
        return dist_cor
    
    def compute_lat_cor(self, coef_number = None, cor_else = True, **kwargs):
        lat_cor = self._compute_coef(self.latitude, coef_number, **kwargs)
        A1 = kwargs['A1']
        A3 = kwargs['A3']
        F = kwargs['F']
        L = kwargs['L']
        m2 = kwargs['m2']
        
        
        if cor_else:
            lat_cor += (-2235*sin(L) + 382*sin(A3) + 175*sin(A1-F) + 175*sin(A1+F) +
                            127*sin(L-m2)-115*sin(L+m2) )
        
        return lat_cor
    
    def compute_moon_coord(self, T, coef_number = None, cor_else = True):
        L = self.mean_longitude(T)
        D = self.mean_elongation(T)
        M = self.sun_mean_anomaly(T)
        m2 = self.moon_mean_anomaly(T)
        F = self.moon_latitude_arguments(T)
        E = self.compute_E(T)
        
        A1 = self.A1_func(T)
        A2 = self.A2_func(T)
        A3 = self.A3_func(T)
        
        
        lat_cor = self.compute_lat_cor(coef_number, cor_else, L = L, D = D, M = M, m2 = m2, F = F, E = E,
                                      A1 = A1, A2 = A2, A3 = A3)
        long_cor = self.compute_long_cor(coef_number, cor_else, L = L, D = D, M = M, m2 = m2, F = F, E = E,
                                      A1 = A1, A2 = A2, A3 = A3)
        dist_cor = self.compute_dist_cor(coef_number, L = L, D = D, M = M, m2 = m2, F = F, E = E,
                                      A1 = A1, A2 = A2, A3 = A3)
        
        lambda_ = L + long_cor/1e6
        beta_ = lat_cor/1e6
        delta_ = 385000.56 + dist_cor/1e3
        
        return lambda_, beta_, delta_

T = julianCalc.compute_T(12, 4, 1992)
mP = moonPosition()
print(mP.compute_moon_coord(T))

File: moonPos.mjs
import { on } from 'events';
import fs from 'fs';
import { julianCalc } from './julianCalc.mjs';
import { customTrig } from './radianTrig.mjs';

const data = JSON.parse(fs.readFileSync('./package/coefficients.json', 'utf-8'));
const longitudeConst = data.moonPosition.longitude;
const latitudeConst = data.moonPosition.latitude;
const distanceConst = data.moonPosition.distance;

const moonConst ={
    meanLongitude(T){
        let L = 218.3164477 + 481267.88123431*T 
                - 0.0015786*T**2 + T**3/538841 
                - T**4/65194000
        
        L = customTrig.reduceAngle(L);
        return L;
    },
    meanElongation(T){
        let D = 297.8501921 + 445267.1114034*T 
                - 0.0018819*T**2 + T**3/545868 
                - T**4/113065000
        
        D = customTrig.reduceAngle(D);
        return D;
    },
    meanSunAnomaly(T){
        let mS = 357.5291092 + 35999.0502909*T 
                - 0.0001536*T**2 + T**3/24490000
        
        mS = customTrig.reduceAngle(mS);
        return mS;
    },
    meanMoonAnomaly(T){
        let mM = 134.9633964 + 477198.8675055*T 
                + 0.0087414*T**2 + T**3/69699 
                - T**4/14712000
        
        mM = customTrig.reduceAngle(mM);
        return mM;
    },
    moonLatitudeArgument(T){
        let F = 93.2720950 + 483202.0175233*T 
                - 0.0036539*T**2 - T**3/3526000 
                + T**4/863310000
        
        F = customTrig.reduceAngle(F);
        return F;
    },
    funcA1(T){
        let A1 = 119.75 + 131.849*T;
        A1 = customTrig.reduceAngle(A1);
        return A1;
    },
    funcA2(T){
        let A2 = 53.09 + 479264.290*T;
        A2 = customTrig.reduceAngle(A2);
        return A2;
    },
    funcA3(T){
        let A3 = 313.45 + 481266.484*T;
        A3 = customTrig.reduceAngle(A3);
        return A3;
    },
    eccentricityFactor(T){
        const E = 1 - 0.002516*T - 0.0000074*T**2;
        return E;
    },
    computeCoef(T, coefDataset, coefNumber = null, useSin = true){
        const D = this.meanElongation(T);
        const mS = this.meanSunAnomaly(T);
        const mM = this.meanMoonAnomaly(T);
        const F = this.moonLatitudeArgument(T);
        const E = this.eccentricityFactor(T);

        let sumAll = 0

        const terms = coefNumber || coefDataset.length;
        const func = useSin ? customTrig.sin : customTrig.cos;

        for (const coef of coefDataset.slice(0, terms)){
            const param = coef[0]*D+coef[1]*mS+coef[2]*mM+coef[3]*F
            let oneTerm = coef[4]*func(param);

            if(Math.abs(coef[1] === 1)){
                oneTerm *= E;
            } else if (Math.abs(coef[1] === 2)){
                oneTerm *= E**2;
            }
            sumAll += oneTerm;
        }
        return sumAll;
    }
};

const moonPos = {
    computeLongitude(T, coefNumber = null){
        const A1 = moonConst.funcA1(T);
        const A2 = moonConst.funcA2(T);
        const F = moonConst.moonLatitudeArgument(T);

        let lambda = moonConst.meanLongitude(T);
        let correction = moonConst.computeCoef(T, data.moonPosition.longitude, coefNumber);
        correction += 3958*customTrig.sin(A1) + 1962*customTrig.sin(lambda-F) + 318*customTrig.sin(A2);
        
        lambda += correction/1e6;
        return lambda;
    },
    computeLatitude(T, coefNumber = null){
        let beta = moonConst.computeCoef(T, data.moonPosition.latitude, coefNumber);
        const A1 = moonConst.funcA1(T);
        const A3 = moonConst.funcA3(T);
        const F = moonConst.moonLatitudeArgument(T);
        const L = moonConst.meanLongitude(T);
        const mM = moonConst.meanMoonAnomaly(T);

        beta += -2235*customTrig.sin(L)+ 382*customTrig.sin(A3)
                +175*customTrig.sin(A1-F) + 175*customTrig.sin(A1+F)
                +127*customTrig.sin(L-mM) - 115*customTrig.sin(L+mM);
        
        return beta/1e6;
    },
    computeDistance(T, coefNumber = null){
        let delta = moonConst.computeCoef(T, data.moonPosition.distance, coefNumber, false);
        delta = 385000.56 + delta/1e3;
        return delta;
    }
}

export { moonConst, moonPos };